harness/determined/_info.py

```diff
@@ -0,0 +1,318 @@
+        self.slot_ids = slot_ids
+        self.task_id = task_id
+        self.allocation_id = allocation_id
+        self.session_token = session_token
+        self.task_type = task_type
+
+        self.master_cert_name = master_cert_name
+        self.master_cert_file = master_cert_file
+        self._latest_checkpoint = latest_checkpoint
+
+        self._trial_info = trial_info
+        self._rendezvous_info = rendezvous_info
+        self._resources_info = resources_info
+
+    @classmethod
+    def _from_env(cls) -> "ClusterInfo":
+        required = [
+            "DET_MASTER",
+            "DET_MASTER_ID",
+            "DET_CLUSTER_ID",
+            "DET_AGENT_ID",
+            "DET_CONTAINER_ID",
+            "DET_SLOT_IDS",
+            "DET_TASK_ID",
+            "DET_ALLOCATION_ID",
+            "DET_SESSION_TOKEN",
+            "DET_TASK_TYPE",
+        ]
+        missing = [r for r in required if r not in os.environ]
+        if missing:
+            raise RuntimeError(
+                f'missing environment keys [{", ".join(missing)}], is this running on-cluster?'
+            )
+        return cls(
+            master_url=os.environ["DET_MASTER"],
+            master_id=os.environ["DET_MASTER_ID"],
+            cluster_id=os.environ["DET_CLUSTER_ID"],
+            agent_id=os.environ["DET_AGENT_ID"],
+            container_id=os.environ["DET_CONTAINER_ID"],
+            slot_ids=json.loads(os.environ["DET_SLOT_IDS"]),
+            task_id=os.environ["DET_TASK_ID"],
+            allocation_id=os.environ["DET_ALLOCATION_ID"],
+            session_token=os.environ["DET_SESSION_TOKEN"],
+            task_type=os.environ["DET_TASK_TYPE"],
+            # Optional info:
+            master_cert_name=os.environ.get("DET_MASTER_CERT_NAME"),
+            master_cert_file=os.environ.get("DET_MASTER_CERT_FILE"),
+            latest_checkpoint=os.environ.get("DET_LATEST_CHECKPOINT"),
+            # Separate info objects:
+            trial_info=TrialInfo._from_file(),
+            rendezvous_info=RendezvousInfo._from_file(),
+            resources_info=ResourcesInfo._from_file(),
+        )
+
+    def _to_file(self, path: str = "/run/determined/info/cluster.json") -> None:
+        skip = ("_trial_info", "_rendezvous_info", "_resources_info")
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items() if k not in skip}, f)
+
+    @classmethod
+    def _from_file(cls, path: str = "/run/determined/info/cluster.json") -> Optional["ClusterInfo"]:
+        if not os.path.exists(path):
+            return None
+        with open(path) as f:
+            return cls(
+                trial_info=TrialInfo._from_file(),
+                rendezvous_info=RendezvousInfo._from_file(),
+                resources_info=ResourcesInfo._from_file(),
+                **json.load(f),
+            )
+
+    @property
+    def latest_checkpoint(self) -> Optional[str]:
+        if self.task_type != "TRIAL":
```

Comment:
Mixed.  On the one hand, it's probably perfectly fine.

On the other hand, "Enum" implies that there's a fixed set that you can reasonably expect to never see values outside of.  But eventually we want to version the master-harness APIs such that the master can be updated and old user images (aka old harness versions) still run.  And in that world, it's perfectly reasonable that the `info.task_type` might contain unknown values, written by a master that is a version ahead of the harness code that is executing.

I think there's a fancy explanation for this: https://stackoverflow.com/a/4605019

A more accurate pattern would be for task_type to be a class, and the trial_task_type to be a subclass of task_type or something.  Because you can always have new subclasses; it's not normal for code to try to enumerate all possible subclasses of PyTorchTrial.  But that is the wrong pattern for other reasons...

Reply:
i'd never thought of that. good read, thanks.